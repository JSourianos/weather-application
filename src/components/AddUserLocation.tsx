// TODO: Test error state
import { useStoredUrls } from "@/src/context/stored-urls";
import { useAsyncStorage } from "@react-native-async-storage/async-storage";
import { useUserLocation } from "@/src/hooks/useUserLocation";
import { Pressable, Text } from "react-native";
import { Navigation } from "lucide-react-native";
import { ASYNC_STORAGE_KEYS } from "@/src/utils/async-storage-keys";

export const AddUserLocation = () => {
  const { addStoredUrl } = useStoredUrls();
  const { setItem } = useAsyncStorage(ASYNC_STORAGE_KEYS.MY_LOCATION);
  const { getCurrentLocation, error: userLocationError } = useUserLocation();

  async function handleAddLocation() {
    const location = await getCurrentLocation();

    if (!location) {
      return;
    }

    const { latitude, longitude } = location;
    const stringifiedLocation = JSON.stringify({ latitude, longitude });

    await setItem(stringifiedLocation, (error) => {
      if (error) {
        console.error(error);
      }

      addStoredUrl({
        id: Math.floor(Math.random() * 1000).toString(),
        name: "My Location",
        url: `https://api.met.no/weatherapi/locationforecast/2.0/compact?lat=${latitude}&lon=${longitude}`,
      });
    });
  }

  if (userLocationError) {
    return <Text className="text-red-500">{userLocationError}</Text>;
  }

  return (
    <Pressable
      className="bg-slate-200 flex flex-row items-center gap-4 p-4 self-start rounded-2xl my-2 ml-auto"
      onPress={handleAddLocation}
    >
      <Navigation />
      <Text className="text-blue-500 font-medium">Add my location!</Text>
    </Pressable>
  );
};
