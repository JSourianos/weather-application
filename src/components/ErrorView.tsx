import { Text, View } from "react-native";

export const ErrorView = () => {
  return (
    <View>
      <Text>Oops, something went wrong...</Text>
    </View>
  );
};
