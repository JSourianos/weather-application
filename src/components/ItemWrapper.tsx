import { View } from "react-native";

export const ItemWrapper = ({ children }: { children: React.ReactNode }) => (
  <View className="bg-slate-200 rounded-2xl h-52 w-full">{children}</View>
);
