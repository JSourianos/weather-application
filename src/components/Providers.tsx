import { ReactNode } from "react";
import { UserPreferenceProvider } from "@/src/context/user-preference";
import { StoredUrlsProvider } from "@/src/context/stored-urls";

export const ApplicationProviders = ({ children }: { children: ReactNode }) => {
  return (
    <UserPreferenceProvider>
      <StoredUrlsProvider>{children}</StoredUrlsProvider>
    </UserPreferenceProvider>
  );
};
