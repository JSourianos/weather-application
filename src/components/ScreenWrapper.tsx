import { SafeAreaView, View } from "react-native";
import { ReactNode } from "react";

export const ScreenWrapper = ({ children }: { children: ReactNode }) => {
  return (
    <SafeAreaView className="bg-white h-full" style={{ flex: 1 }}>
      <View className="px-4" style={{ flex: 1 }}>
        {children}
      </View>
    </SafeAreaView>
  );
};
