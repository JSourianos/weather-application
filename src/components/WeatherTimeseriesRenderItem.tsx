import { YRTimeseries } from "@/src/lib/utils/types";
import { useUserPreference } from "@/src/context/user-preference";
import dayjs from "dayjs";
import { Text, View } from "react-native";
import { DropletsIcon, ThermometerIcon, WindIcon } from "@/src/lib/icons";
import { resolveDegreeSelectionText } from "@/src/lib/utils/utils";

type WeatherTimeseriesRenderItemProps = {
  item: YRTimeseries;
};

export const WeatherTimeseriesRenderItem = ({
  item,
}: WeatherTimeseriesRenderItemProps) => {
  const { degreeType } = useUserPreference();
  const details = item.data.instant.details;
  const time = dayjs(item.time).format("HH:mm");
  return (
    <View className="p-4 bg-slate-50 rounded-2xl border-2 border-slate-300">
      <Text className="text-3xl font-medium mb-4">{time}</Text>
      <View className="flex flex-row justify-between items-center">
        <View className="flex items-center gap-2">
          <WindIcon className="h-5 w-5 text-blue-500" />
          <Text className="font-medium">{details.wind_speed} km/h</Text>
        </View>
        <View className="flex items-center gap-2">
          <ThermometerIcon className="h-5 w-5 text-blue-500" />
          <Text className="font-medium">
            {resolveDegreeSelectionText(details.air_temperature, degreeType)}
          </Text>
        </View>
        <View className="flex items-center gap-2">
          <DropletsIcon className="h-5 w-5 text-blue-500" />
          <Text className="font-medium">{details.relative_humidity}%</Text>
        </View>
      </View>
    </View>
  );
};
