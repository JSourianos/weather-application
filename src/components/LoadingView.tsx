import { ActivityIndicator, Text } from "react-native";
import { ScreenWrapper } from "@/src/components/ScreenWrapper";

export const LoadingView = () => {
  return (
    <ScreenWrapper>
      <ActivityIndicator size="large" />
      <Text>Loading...</Text>
    </ScreenWrapper>
  );
};
