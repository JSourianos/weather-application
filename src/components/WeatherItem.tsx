import { View, Text, Pressable } from "react-native";
import { useFetchWeatherData } from "@/src/hooks/useFetchWeatherData";
import { ErrorView } from "@/src/components/ErrorView";
import { LoadingView } from "@/src/components/LoadingView";
import { Cloud, Droplets, Wind } from "lucide-react-native";
import { Link } from "expo-router";
import { resolveDegreeSelectionText } from "@/src/utils/utils";
import { useUserPreference } from "@/src/context/user-preference";

export const WeatherItem = ({
  url,
  name,
  id,
}: {
  url: string;
  name: string;
  id: string;
}) => {
  const { data, status } = useFetchWeatherData(url);
  const { degreeType } = useUserPreference();

  if (status === "error" || data === null) {
    return <ErrorView />;
  }

  if (status === "loading") {
    return <LoadingView />;
  }

  const { air_temperature, relative_humidity, wind_speed } =
    data.properties.timeseries[0].data.instant.details;

  return (
    <Link href={`/weather/${id}`} asChild>
      <Pressable>
        <BaseContainer>
          <Text className="text-2xl font-bold text-left p-4">{name}</Text>
          <View className="flex items-center justify-center mb-6">
            <View className="col-span-2 flex items-center justify-center">
              <Cloud className="h-12 w-12 text-blue-500 mr-4" />
              <Text className="text-4xl font-bold">
                {" "}
                {resolveDegreeSelectionText(air_temperature, degreeType)}
              </Text>
            </View>
            <View className="flex flex-row justify-center gap-8  mt-4 w-full">
              <View className="flex items-center">
                <Droplets className="h-5 w-5 text-blue-500 mr-2" />
                <Text>Humidity: {relative_humidity}%</Text>
              </View>
              <View className="flex items-center">
                <Wind className="h-5 w-5 text-blue-500 mr-2" />

                <Text>Wind: {wind_speed} km/h</Text>
              </View>
            </View>
          </View>
        </BaseContainer>
      </Pressable>
    </Link>
  );
};

const BaseContainer = ({ children }: { children: React.ReactNode }) => {
  return (
    <View className="bg-blue-50 rounded-2xl border border-blue-200 ">
      {children}
    </View>
  );
};
