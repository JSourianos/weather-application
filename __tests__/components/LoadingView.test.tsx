import { render } from "@testing-library/react-native";
import { LoadingView } from "@/src/components/LoadingView";

describe("LoadingviEW component", () => {
  it("matches the snapshot", () => {
    const { toJSON } = render(<LoadingView />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders the correct loading message", () => {
    const { getByText } = render(<LoadingView />);
    expect(getByText("Loading...")).toBeTruthy();
  });
});
