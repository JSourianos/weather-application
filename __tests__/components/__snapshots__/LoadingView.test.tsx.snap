// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingviEW component matches the snapshot 1`] = `
<RCTSafeAreaView
  className="bg-white h-full"
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    className="px-4"
    style={
      {
        "flex": 1,
      }
    }
  >
    <ActivityIndicator
      size="large"
    />
    <Text>
      Loading...
    </Text>
  </View>
</RCTSafeAreaView>
`;
